src/metadata_map.rs

```diff
@@ -0,0 +1,1716 @@
+    ///     println!("{:?}", value);
+    /// }
+    /// ```
+    pub fn values(&self) -> Values<MetadataValue> {
+        Values { inner: self.headers.values() }
+    }
+
+/*
+    /// An iterator visiting all values mutably.
+    ///
+    /// The iteration order is arbitrary, but consistent across platforms for
+    /// the same crate version.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let mut map = MetadataMap::default();
+    ///
+    /// map.insert("x-word", "hello".parse().unwrap());
+    /// map.append("x-word", "goodbye".parse().unwrap());
+    /// map.insert("x-number", "123".parse().unwrap());
+    ///
+    /// for value in map.values_mut() {
+    ///     value.set_sensitive(true);
+    /// }
+    /// ```
+    // TODO(pgron): Expose this. Blocked by https://github.com/hyperium/http/pull/278
+    pub fn values_mut(&mut self) -> http::header::ValuesMut<MetadataValue> {
+        self.headers.values_mut()
+    }*/
+
+    /// Clears the map, returning all entries as an iterator.
+    ///
+    /// The internal memory is kept for reuse.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let mut map = MetadataMap::new();
+    ///
+    /// map.insert("x-word", "hello".parse().unwrap());
+    /// map.append("x-word", "goodbye".parse().unwrap());
+    /// map.insert("x-number", "123".parse().unwrap());
+    ///
+    /// let mut drain = map.drain();
+    ///
+    /// let (key, mut vals) = drain.next().unwrap();
+    ///
+    /// assert_eq!("x-word", key);
+    /// assert_eq!("hello", vals.next().unwrap());
+    /// assert_eq!("goodbye", vals.next().unwrap());
+    /// assert!(vals.next().is_none());
+    ///
+    /// let (key, mut vals) = drain.next().unwrap();
+    ///
+    /// assert_eq!("x-number", key);
+    /// assert_eq!("123", vals.next().unwrap());
+    /// assert!(vals.next().is_none());
+    /// ```
+    pub fn drain(&mut self) -> Drain<MetadataValue> {
+        Drain { inner: self.headers.drain() }
+    }
+
+    /// Gets the given key's corresponding entry in the map for in-place
+    /// manipulation.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let mut map = MetadataMap::default();
+    ///
+    /// let headers = &[
+    ///     "content-length",
+    ///     "x-hello",
+    ///     "Content-Length",
+    ///     "x-world",
+    /// ];
+    ///
+    /// for &header in headers {
+    ///     let counter = map.entry(header).unwrap().or_insert("".parse().unwrap());
+    ///     *counter = format!("{}{}", counter.to_str().unwrap(), "1").parse().unwrap();
+    /// }
+    ///
+    /// assert_eq!(map.get("content-length").unwrap(), "11");
+    /// assert_eq!(map.get("x-hello").unwrap(), "1");
+    /// ```
+    pub fn entry<K>(&mut self, key: K) -> Result<Entry<MetadataValue>, InvalidMetadataKey>
+        where K: AsMetadataKey
+    {
+        match key.entry(self) {
```

Comment:
Can this be implemented in terms of `HeaderMap::entry`?

Reply:
I'm too much of a Rust noob to be able to answer that question unfortunately. I want users of this function to be able to pass in `MetadataKey`s to this method. But `http::header::AsHeaderName` is a sealed trait so as far as I know I can't just add support for `MetadataKey` to it and replace `key.entry(self)` here with `self.headers.entry(key)` and make things work from there.

What did you have in mind?